### PR TITLE
OCPBUGSM-1928868 : Add validation for Auto-assign role

### PR DESCRIPTION
**Pull request by rewantsoni**
Signed off by: Rewant Soni resoni@redhat.com

### CODE REVIEW

File: internal/operators/ocs/validations.go [hunk 1]
```diff
@@ -110,6 +110,20 @@ func (o *ocsValidator) ValidateOCSRequirements(cluster *models.Cluster) bool {
 		return false
 	} else {
 		for _, host := range hosts { // if the worker nodes >=3 , install OCS on all the worker nodes if they satisfy OCS requirements
+
+			// if number of hosts are >=6 and one of them has Role as Auto-assign
+			if host.Role == models.HostRoleAutoAssign {
+				status = "Not supporting OCS installation if any 1 host has role Auto-assign for number of host >=6"
```
Comment (omertuc) [inline internal/operators/ocs/validations.go:116]: @avishayt user facing message, might want to take a look

File: internal/operators/ocs/validations.go [hunk 2]
```diff
@@ -110,6 +110,20 @@ func (o *ocsValidator) ValidateOCSRequirements(cluster *models.Cluster) bool {
 		return false
 	} else {
 		for _, host := range hosts { // if the worker nodes >=3 , install OCS on all the worker nodes if they satisfy OCS requirements
+
+			// if number of hosts are >=6 and one of them has Role as Auto-assign
```
Comment (omertuc) [inline internal/operators/ocs/validations.go:114]: Can the comment elaborate on why this is a limitation?

File: internal/operators/ocs/validations.go [hunk 3]
```diff
@@ -110,6 +110,20 @@ func (o *ocsValidator) ValidateOCSRequirements(cluster *models.Cluster) bool {
...
+			if host.Role == models.HostRoleAutoAssign {
+				status = "Not supporting OCS installation if any 1 host has role Auto-assign for number of host >=6"
+				o.log.Info(status)
+				o.log.Info("Setting Operator Status")
+				err = setOperatorStatus(cluster, status)
+				if err != nil {
+					o.log.Error("Failed to set Operator status", err)
+					return false
+				}
+				o.log.Info(status)
```
Comment (gamli75) [inline internal/operators/ocs/validations.go:124]: duplicate of line 117 - I don't understand why we need it

File: internal/operators/ocs/validations.go [hunk 4]
```diff
@@ -110,6 +110,20 @@ func (o *ocsValidator) ValidateOCSRequirements(cluster *models.Cluster) bool {
 		return false
 	} else {
 		for _, host := range hosts { // if the worker nodes >=3 , install OCS on all the worker nodes if they satisfy OCS requirements
+
+			// if number of hosts are >=6 and one of them has Role as Auto-assign
+			if host.Role == models.HostRoleAutoAssign {
+				status = "Not supporting OCS installation if any 1 host has role Auto-assign for number of host >=6"
+				o.log.Info(status)
+				o.log.Info("Setting Operator Status")
```
Comment (gamli75) [inline internal/operators/ocs/validations.go:118]: Please combine lines 117 & 118 to one logline

File: internal/operators/ocs/validations.go (hunk 2 above)
Comment (rewantsoni) [reply internal/operators/ocs/validations.go]: Done

File: internal/operators/ocs/validations.go (hunk 4 above)
Comment (rewantsoni) [reply internal/operators/ocs/validations.go]: Done

File: internal/operators/ocs/validations.go (hunk 3 above)
Comment (rewantsoni) [reply internal/operators/ocs/validations.go]: Removed

File: internal/operators/ocs/validations.go [hunk 5]
```diff
@@ -110,6 +110,21 @@ func (o *ocsValidator) ValidateOCSRequirements(cluster *models.Cluster) bool {
...
 	} else {
 		for _, host := range hosts { // if the worker nodes >=3 , install OCS on all the worker nodes if they satisfy OCS requirements
+
+			/* if number of hosts are >=6 and one of them has Role as Auto-assign,
+			it is a problem because we won't know that node would be a master or a worker as the role would be Auto assign
+			Hence without knowing that won't be able to run OCS validations
+			*/
+			if host.Role == models.HostRoleAutoAssign {
+				status = "Not supporting OCS installation if any 1 host has role Auto-assign for number of host >=6"
+				o.log.Info(status, "\nSetting Operator Status")
```
Comment (gamli75) [inline internal/operators/ocs/validations.go:120]: please change it to o.log.Info("Setting Operator Status ", status) and please do it for lines https://github.com/openshift/assisted-service/blob/master/internal/operators/ocs/validations.go#L102

File: internal/operators/ocs/validations.go (hunk 1 above)
Comment (avishayt) [reply internal/operators/ocs/validations.go:118]: Thanks @omertuc 

I prefer: "All host roles must be assigned to enable OCS"

Comment (rewantsoni) [reply internal/operators/ocs/validations.go]: Done

File: internal/cluster/transition_test.go [hunk 6]
```diff
@@ -4220,6 +4220,69 @@ var _ = Describe("Ocs Operator use-cases", func() {
...
+				isApiVipValid:                       {status: ValidationSuccess, messagePattern: "belongs to the Machine CIDR and is not in use."},
+				isIngressVipDefined:                 {status: ValidationSuccess, messagePattern: "The Ingress virtual IP is defined"},
+				isIngressVipValid:                   {status: ValidationSuccess, messagePattern: "belongs to the Machine CIDR and is not in use."},
+				AllHostsAreReadyToInstall:           {status: ValidationSuccess, messagePattern: "All hosts in the cluster are ready to install"},
+				IsDNSDomainDefined:                  {status: ValidationSuccess, messagePattern: "The base domain is defined"},
+				IsPullSecretSet:                     {status: ValidationSuccess, messagePattern: "The pull secret is set"},
+				SufficientMastersCount:              {status: ValidationSuccess, messagePattern: "The cluster has a sufficient number of master candidates."},
+				IsOcsRequirementsSatisfied:          {status: ValidationFailure, messagePattern: "Not supporting OCS installation if any 1 host has role Auto-assign for number of host >=6"},
```
Comment (priyanka19-98) [inline internal/cluster/transition_test.go:4252]: Please correct the failure message.

File: internal/cluster/transition_test.go [hunk 7]
```diff
@@ -4214,6 +4216,69 @@ var _ = Describe("Ocs Operator use-cases", func() {
...
+				IsPullSecretSet:                     {status: ValidationSuccess, messagePattern: "The pull secret is set"},
+				SufficientMastersCount:              {status: ValidationSuccess, messagePattern: "The cluster has a sufficient number of master candidates."},
+				IsOcsRequirementsSatisfied:          {status: ValidationFailure, messagePattern: "All host roles must be assigned to enable OCS."},
+			}),
+			errorExpected: false,
+		},
+		{
+			name:               "ocs enabled, 3 nodes, with role of one as auto-assign",
+			srcState:           models.ClusterStatusPendingForInput,
+			dstState:           models.ClusterStatusReady,
```
Comment (priyanka19-98) [inline internal/cluster/transition_test.go:4255]: @gamli75 , could you please verify the transition states for `Auto Assign` mode?

File: internal/cluster/transition_test.go [hunk 8]
```diff
@@ -4242,12 +4307,13 @@ var _ = Describe("Ocs Operator use-cases", func() {
...
 
 			Expect(db.Create(&cluster).Error).ShouldNot(HaveOccurred())
+			mockIsValidMasterCandidate()
 			for i := range t.hosts {
 				t.hosts[i].ClusterID = clusterId
 				Expect(db.Create(&t.hosts[i]).Error).ShouldNot(HaveOccurred())
 			}
 			cluster = getCluster(clusterId, db)
-			if t.dstState == models.ClusterStatusInsufficient {
+			if t.dstState == models.ClusterStatusInsufficient || t.name == "ocs enabled, 3 nodes, with role of one as auto-assign" {
```
Comment (priyanka19-98) [inline internal/cluster/transition_test.go:4316]: Why do we need to add the extra condition here?

Comment (rewantsoni) [reply internal/cluster/transition_test.go:4316]: Done

**Comment by priyanka19-98** [comment]
@rewantsoni , could you squash your commits to one commit?

File: internal/cluster/transition_test.go [hunk 9]
```diff
@@ -4214,6 +4216,69 @@ var _ = Describe("Ocs Operator use-cases", func() {
...
+			name:               "ocs enabled, 6 nodes, with role of one as auto-assign",
+			srcState:           models.ClusterStatusPendingForInput,
+			dstState:           models.ClusterStatusInsufficient,
+			machineNetworkCidr: "1.2.3.0/24",
+			apiVip:             "1.2.3.5",
+			ingressVip:         "1.2.3.6",
+			dnsDomain:          "test.com",
+			pullSecretSet:      true,
+			hosts: []models.Host{
+				{ID: &hid1, Status: swag.String(models.HostStatusKnown), Inventory: defaultInventory(), Role: models.HostRoleAutoAssign},
```
Comment (nbalacha) [inline internal/cluster/transition_test.go:4229]: Will the list of hosts be returned in the same order in which they were created ? IIf not, how many CPUs/RAM does the default inventory have?

File: internal/cluster/transition_test.go (hunk 7 above)
Comment (gamli75) [reply internal/cluster/transition_test.go:4255]: the dstState should be insufficient and we should remove the condition " || t.name == "ocs enabled, 3 nodes, with role of one as auto-assign"  from line 4316

File: internal/cluster/transition_test.go [hunk 10]
```diff
@@ -4214,6 +4216,69 @@ var _ = Describe("Ocs Operator use-cases", func() {
...
+				isApiVipDefined:                     {status: ValidationSuccess, messagePattern: "The API virtual IP is defined"},
+				isApiVipValid:                       {status: ValidationSuccess, messagePattern: "belongs to the Machine CIDR and is not in use."},
+				isIngressVipDefined:                 {status: ValidationSuccess, messagePattern: "The Ingress virtual IP is defined"},
+				isIngressVipValid:                   {status: ValidationSuccess, messagePattern: "belongs to the Machine CIDR and is not in use."},
+				AllHostsAreReadyToInstall:           {status: ValidationSuccess, messagePattern: "All hosts in the cluster are ready to install"},
+				IsDNSDomainDefined:                  {status: ValidationSuccess, messagePattern: "The base domain is defined"},
+				IsPullSecretSet:                     {status: ValidationSuccess, messagePattern: "The pull secret is set"},
+				SufficientMastersCount:              {status: ValidationSuccess, messagePattern: "The cluster has a sufficient number of master candidates."},
+				IsOcsRequirementsSatisfied:          {status: ValidationSuccess, messagePattern: "OCS Requirements for Compact Mode are satisfied"},
```
Comment (gamli75) [inline internal/cluster/transition_test.go:4278]: the validation status should be Failure and not a success

File: internal/cluster/transition_test.go [hunk 11]
```diff
@@ -3640,7 +3640,9 @@ var _ = Describe("Ocs Operator use-cases", func() {
 		ctrl                                          *gomock.Controller
 		dbName                                        string = "cluster_transition_test_with_ocs_validations"
 	)
-
+	mockIsValidMasterCandidate := func() {
```
Comment (nbalacha) [inline internal/cluster/transition_test.go:3643]: Why is this required?

File: internal/cluster/transition_test.go (hunk 10 above)
Comment (priyanka19-98) [reply internal/cluster/transition_test.go:4278]: I remember from the discussion, that if only 3 hosts are present, the `HostRole` doesn't matter as anyway all the three hosts will be Masters.

Comment (gamli75) [reply internal/cluster/transition_test.go:4278]: you are right - I was confusing between the 2 tests. I think that we should change the test name so that it will be more clear - which test is the negative case(ocs validation failure) and positive case(ocs validation success)

File: internal/cluster/transition_test.go (hunk 7 above)
Comment (gamli75) [reply internal/cluster/transition_test.go:4255]: yes, it's right

File: internal/cluster/transition_test.go (hunk 10 above)
Comment (priyanka19-98) [reply internal/cluster/transition_test.go:4278]: @rewantsoni ^^

**Review by nbalacha — CHANGES_REQUESTED** [review]

File: internal/operators/ocs/validations.go [hunk 12]
```diff
@@ -110,6 +109,21 @@ func (o *ocsValidator) ValidateOCSRequirements(cluster *models.Cluster) string {
...
+			it is a problem because we won't know that node would be a master or a worker as the role would be Auto assign
+			Hence without knowing that won't be able to run OCS validations
+			*/
+			if host.Role == models.HostRoleAutoAssign {
+				status = "All host roles must be assigned to enable OCS."
+				o.log.Info("Setting Operator Status ", status)
+				err = setOperatorStatus(cluster, status)
+				if err != nil {
+					o.log.Error("Failed to set Operator status", err)
+					return validationFailure
```
[the remark this answers is not in the document]
Comment (rewantsoni) [reply internal/operators/ocs/validations.go:123]: Done

File: internal/cluster/transition_test.go (hunk 11 above)
Comment (rewantsoni) [reply internal/cluster/transition_test.go:3643]: If the role is autoassign we need to mock the IsValidMasterCandidate as defined in the interface

File: internal/operators/ocs/validations.go [hunk 13]
```diff
@@ -110,6 +109,21 @@ func (o *ocsValidator) ValidateOCSRequirements(cluster *models.Cluster) string {
 		return validationFailure
 	} else {
 		for _, host := range hosts { // if the worker nodes >=3 , install OCS on all the worker nodes if they satisfy OCS requirements
+
+			/* if number of hosts are >=6 and one of them has Role as Auto-assign,
```
[the remark this answers is not in the document]
Comment (rewantsoni) [reply internal/operators/ocs/validations.go:113]: Done

File: internal/cluster/transition_test.go (hunk 10 above)
Comment (rewantsoni) [reply internal/cluster/transition_test.go:4278]: added

File: internal/cluster/transition_test.go (hunk 9 above)
Comment (rewantsoni) [reply internal/cluster/transition_test.go:4229]: Its there in https://github.com/openshift/assisted-service/blob/0372c7aeb81112b92f2c59ee9b2c5e8abfa4170a/internal/cluster/cluster_test.go#L1168
